container/kvm/sync.go

```diff
@@ -0,0 +1,255 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// +build linux
+// +build amd64 arm64 ppc64el
+
+package kvm
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/series"
+
+	"github.com/juju/juju/environs/imagedownloads"
+	"github.com/juju/juju/environs/simplestreams"
+	"github.com/juju/juju/juju/paths"
+)
+
+const (
+	// FType is the file type we want to fetch and use for kvm instances.
+	FType = "disk1.img"
+)
+
+// Oner is an interface which allows us to use sync params to call
+// imagedownloads.One or pass in a fake for testing.
+type Oner interface {
+	One() (*imagedownloads.Metadata, error)
+}
+
+// syncParams conveys the information necessary for calling imagedownloads.One.
+type syncParams struct {
+	arch, series, ftype string
+	srcFunc             func() simplestreams.DataSource
```

Comment:
It is a func because we allow passing in a custom simplestreams url (if you make your own golden images/tools/whatever) and building a datasource from that. If the arg is empty then the default source is used. This was an early bit of work from the simplestreams implementation, so it isn't fresh in my head. I'll look at it after I get through the rest of the review.